2021/10/15/string-number-conversion.md

```diff
@@ -0,0 +1,64 @@
+---
+author: "Jeff Laughlin"
+title: "On the Importance of Explicitly Converting Strings to Numbers"
+tags: programming, javascript, typescript, types, errors, debugging
+gh_issue_number: 1780
+---
+
+Recently a valued colleague approached me with a JavaScript problem. This individual is new to programming and is working on a self-taught course. The assignment was fairly simple; take a list of space delimited integers and find the maximum and minimum values. If you are an experienced developer you can probably already guess where this is going.
+
+```javascript
+function highAndLow(numbers){
+    const myArr = numbers.split(" ")
+    console.log(myArr);
+    let lowNum = myArr[0];
+    let highNum = myArr[0];
+    for (let i = 0; i<myArr.length; i++) {
+      if (myArr[i] > highNum) {
+        highNum = myArr[i];
+      } else if (myArr[i] < lowNum) {
+        lowNum = myArr[i];
+      }
+      console.log(highNum + ' ' + lowNum);
+    }
+    return highNum + ' ' + lowNum
+  }
+
+console.log(highAndLow("8 3 -5 42 -1 0 0 -9 4 7 4 -4"))
+```
+
+and this produced the output
+
+```javascript
+"8 -1"
+```
+
+These are clearly not the maximum or minimum values.
+
+After looking at it for a few moments I recognized a classic JavaScript pitfall; “failure to explicitly convert stringy numbers to actual number types.”
+
+You see, JavaScript tries to be clever. JavaScript tries to get it right. JavaScript tries to say “the thing you are doing looks like something that you would do with numbers so I’m going to automatically convert these stringy numbers to number-numbers for you.” The problem is that JavaScript is NOT clever, it is in fact very dumb about this. The further problem is when developers come to trust and rely on automatic conversion. Careers have been ruined that way.
+
+In this case the naive programmer would say “Well I’m comparing the things with a mathematical operator (< and >) so JS should treat the values as numbers, right?” Wrong. JavaScript compares them alphabetically, NOT numerically. Except that even the “alphabetical” comparison kind of sucks but that’s another blog post. JS doesn't even attempt to convert to numbers in this case.
+
+Repeat after me: “ALWAYS EXPLICITLY CONVERT STRINGY NUMBERS TO ACTUAL NUMBERS EVEN IF THE LANGUAGE CLAIMS TO DO IT AUTOMATICALLY”. I don’t care if it’s JavaScript, Perl, some fancy Python package, it doesn’t matter. DO NOT TRUST AUTOMATIC TYPE CONVERSION. You will get it wrong. It will get it wrong. There will be tears.
```

Comment:
Perl does not apply here, because it uses different operators for strings vs. numbers. In Perl, `<` and `>` always cause a numeric comparison, so they can be used on `"10"` or `10` identically. And in Perl, `lt` and `gt` are the string comparison operators, which again work the same to show `3` is greater than `"20"`.

Bourne shell (and bash and zsh) likewise have separate operators for strings and numbers, but they're opposite of Perl's: `<` and `>` are for strings, and `-lt` and `-gt` are for numbers.

PHP has yet another way of handling this, using a single set of operators and using a numeric comparison if either type is numeric: https://www.php.net/manual/en/language.operators.comparison.php - which I think is basically exactly as risky as the JavaScript way.

Unless you want to go into those details, probably best to just omit mention of Perl because it just is different from what you're showing here: you can mess it up, by that's by choosing the wrong operator and the operand type doesn't matter. But PHP would be a good one to include IMO.

I'm curious about your mention of "some fancy Python package", since Python's string objects and numbers prevent these kind of mistakes normally, but maybe someone made some package to try to "do what I mean" and ended up importing the JavaScript problem to Python too?!